cluster-api-gcp/cloud/google/machineactuator.go

```diff
@@ -673,3 +694,42 @@ func getSubnet(netRange clusterv1.NetworkRanges) string {
 	}
 	return netRange.CIDRBlocks[0]
 }
+
+// Takes a Machine object representing the current state, and update it based on the actual
+// current state as reported by GCE, kubelet etc.
+func currentMachineActualStatus(gce *GCEClient, currentMachine *clusterv1.Machine) (*clusterv1.Machine, error) {
+	project := currentMachine.ObjectMeta.Annotations[ProjectAnnotationKey]
```

Comment:
Is there a good reason not to reuse existing logic to fetch the current instance? (ie. gce.instanceIfExists) It already handles some cases that this logic overlooks Eg. UID mismatch.

Reply:
Mainly because instanceIfExists uses the providerconfig zone to find the instance. Which won't work correctly if I'm changing zones of my machine (it won't find it machine in the new zone, and now you'll have two actual VMs).

In this implementation we use the annotation that's updated after creating the instance, so changing zones _should_ work as expected.

Reply:
instanceIfExists pulls the zone from the saved status (ie. the last applied state). It should still pull the correct instance even if the zone of the spec on the machine has changed.

As per the code comment here,
> 	// Try to use the last saved status locating the machine
>	// in case instance details like the proj or zone has changed
https://github.com/kubernetes/kube-deploy/blob/master/cluster-api-gcp/cloud/google/machineactuator.go#L484

the instanceIfExists is intended to work in spite of spec zone changes. If it does not correctly fetch the current instance, that needs to be fixed in instanceIfExists.

Reply:
I think you're right. Since we pass currentMachine (which is actually the instance-status of the goal machine), it will use the provider config of the last known machine (aka the currentMachine). I've replaced this call with the existing method.